### ISSUE TITLE
Question about masking & loss on padded EOS tokens during training MLDM model like LLaDA.

### ISSUE THREAD

**Issue by SeaBird-Go** [ZHZisZZ/dllm#81]
Hi, thanks for open-sourcing this repo — I’m trying to understand the data corruption / masking design in training **LLaDA**.

From what I see, each training sequence already ends with a real end-of-turn token (<|eot_id|>). However, within a batch, shorter sequences are right-padded using EOS(`<|endoftext|>`) tokens, and these padded EOS positions **can also be randomly masked and included in the loss** (loss computed only on masked indices). And the **attention_mask** is set to **None** during training.

For example, this is one of the actual padded decoded input_ids:
<img width="2802" height="540" alt="Image" src="https://github.com/user-attachments/assets/ac7d0383-49e9-4664-81cc-b27f5d091b31" />

My questions:

- Is it intentional that right-padded EOS tokens (i.e., padding, not the “real” end token <|eot_id|>) are included in the corruption process and contribute to the training loss?
- What is the motivation behind using EOS as padding here, given that the sequence already ends with <|eot_id|>?
- Wouldn’t this introduce a distribution shift where the model sees many repeated EOS tokens at the end of sequences (and learns to predict them), potentially diluting the learning signal and/or affecting termination behavior at inference?
- Would it be more appropriate to exclude padded positions from masking and/or loss via an attention mask / loss mask (or to use a dedicated PAD token)?

**Comment by SeaBird-Go** [comment]
Thanks for pointing out the `group_by_length` setting. I found it's a rough problem for DLM to correctly predict the EOS to terminate the answer. When I tested the llada by running `example/llada/sample.py`, one of the outputs are:

<img width="1070" height="260" alt="Image" src="https://github.com/user-attachments/assets/6f04f97c-7167-40d6-bc14-73eea9f3c98b" />

However, when I set the `max_new_tokens` = 512, the results tend to incorrect:

<img width="1629" height="526" alt="Image" src="https://github.com/user-attachments/assets/28af4669-be1a-4521-9633-675b320a26c9" />


Do you have more solutions to solve this problem?

BTW, did you find the tokenizer treats the `<|eot_id|>` token differently for `GSAI-ML/LLaDA-8B-Base` and `GSAI-ML/LLaDA-8B-Instruct`? The former is:
<img width="728" height="53" alt="Image" src="https://github.com/user-attachments/assets/6217f80d-10d2-45d5-8a18-182663e0762d" />


While the latter is:

<img width="451" height="101" alt="Image" src="https://github.com/user-attachments/assets/b60d60c8-3da1-4024-b53a-c04b78844743" />

It's so strange.

**Comment by lingjiechen2** [comment]
You’ve pointed out some common issues in current DLMs.

For the EOS prediction issue, this is a well-known challenge across many DLMs. In practice, it can sometimes be mitigated by adjusting decoding hyperparameters such as the temperature or selecting a more appropriate max_new_tokens. There is also ongoing research exploring more robust solutions to this problem.

Regarding the `eot_id` token, this comes from the tokenizer used in the current LLaDA model. We’ve already noted this issue in the codebase (see [dllm/utils/models.py](https://github.com/ZHZisZZ/dllm/blob/main/dllm/utils/models.py#L141)). So far, we haven’t modified the tokenizer to avoid introducing additional inconsistencies and to stay aligned with the original implementation.
